<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Alencar Xavier and Shizhong Xu" />

<meta name="date" content="2016-09-14" />

<title>Meta-analysis</title>






<link href="data:text/css,body%20%7B%0A%20%20background%2Dcolor%3A%20%23fff%3B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20max%2Dwidth%3A%20700px%3B%0A%20%20overflow%3A%20visible%3B%0A%20%20padding%2Dleft%3A%202em%3B%0A%20%20padding%2Dright%3A%202em%3B%0A%20%20font%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0A%20%20font%2Dsize%3A%2014px%3B%0A%20%20line%2Dheight%3A%201%2E35%3B%0A%7D%0A%0A%23header%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0A%0A%23TOC%20%7B%0A%20%20clear%3A%20both%3B%0A%20%20margin%3A%200%200%2010px%2010px%3B%0A%20%20padding%3A%204px%3B%0A%20%20width%3A%20400px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20border%2Dradius%3A%205px%3B%0A%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20font%2Dsize%3A%2013px%3B%0A%20%20line%2Dheight%3A%201%2E3%3B%0A%7D%0A%20%20%23TOC%20%2Etoctitle%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%20%20font%2Dsize%3A%2015px%3B%0A%20%20%20%20margin%2Dleft%3A%205px%3B%0A%20%20%7D%0A%0A%20%20%23TOC%20ul%20%7B%0A%20%20%20%20padding%2Dleft%3A%2040px%3B%0A%20%20%20%20margin%2Dleft%3A%20%2D1%2E5em%3B%0A%20%20%20%20margin%2Dtop%3A%205px%3B%0A%20%20%20%20margin%2Dbottom%3A%205px%3B%0A%20%20%7D%0A%20%20%23TOC%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dleft%3A%20%2D2em%3B%0A%20%20%7D%0A%20%20%23TOC%20li%20%7B%0A%20%20%20%20line%2Dheight%3A%2016px%3B%0A%20%20%7D%0A%0Atable%20%7B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dcolor%3A%20%23DDDDDD%3B%0A%20%20border%2Dstyle%3A%20outset%3B%0A%20%20border%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0A%20%20border%2Dwidth%3A%202px%3B%0A%20%20padding%3A%205px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%20%20line%2Dheight%3A%2018px%3B%0A%20%20padding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0A%20%20border%2Dleft%2Dstyle%3A%20none%3B%0A%20%20border%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Ap%20%7B%0A%20%20margin%3A%200%2E5em%200%3B%0A%7D%0A%0Ablockquote%20%7B%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20padding%3A%200%2E25em%200%2E75em%3B%0A%7D%0A%0Ahr%20%7B%0A%20%20border%2Dstyle%3A%20solid%3B%0A%20%20border%3A%20none%3B%0A%20%20border%2Dtop%3A%201px%20solid%20%23777%3B%0A%20%20margin%3A%2028px%200%3B%0A%7D%0A%0Adl%20%7B%0A%20%20margin%2Dleft%3A%200%3B%0A%7D%0A%20%20dl%20dd%20%7B%0A%20%20%20%20margin%2Dbottom%3A%2013px%3B%0A%20%20%20%20margin%2Dleft%3A%2013px%3B%0A%20%20%7D%0A%20%20dl%20dt%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%7D%0A%0Aul%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%7D%0A%20%20ul%20li%20%7B%0A%20%20%20%20list%2Dstyle%3A%20circle%20outside%3B%0A%20%20%7D%0A%20%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dbottom%3A%200%3B%0A%20%20%7D%0A%0Apre%2C%20code%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20color%3A%20%23333%3B%0A%20%20white%2Dspace%3A%20pre%2Dwrap%3B%20%20%20%20%2F%2A%20Wrap%20long%20lines%20%2A%2F%0A%7D%0Apre%20%7B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20margin%3A%205px%200px%2010px%200px%3B%0A%20%20padding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Acode%20%7B%0A%20%20font%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0A%20%20font%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0A%20%20padding%3A%202px%200px%3B%0A%7D%0A%0Adiv%2Efigure%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0A%20%20background%2Dcolor%3A%20%23FFFFFF%3B%0A%20%20padding%3A%202px%3B%0A%20%20border%3A%201px%20solid%20%23DDDDDD%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20margin%3A%200%205px%3B%0A%7D%0A%0Ah1%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%20%20font%2Dsize%3A%2035px%3B%0A%20%20line%2Dheight%3A%2040px%3B%0A%7D%0A%0Ah2%20%7B%0A%20%20border%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20padding%2Dbottom%3A%202px%3B%0A%20%20font%2Dsize%3A%20145%25%3B%0A%7D%0A%0Ah3%20%7B%0A%20%20border%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20font%2Dsize%3A%20120%25%3B%0A%7D%0A%0Ah4%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0A%20%20margin%2Dleft%3A%208px%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Ah5%2C%20h6%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23ccc%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Aa%20%7B%0A%20%20color%3A%20%230033dd%3B%0A%20%20text%2Ddecoration%3A%20none%3B%0A%7D%0A%20%20a%3Ahover%20%7B%0A%20%20%20%20color%3A%20%236666ff%3B%20%7D%0A%20%20a%3Avisited%20%7B%0A%20%20%20%20color%3A%20%23800080%3B%20%7D%0A%20%20a%3Avisited%3Ahover%20%7B%0A%20%20%20%20color%3A%20%23BB00BB%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%0A%2F%2A%20Class%20described%20in%20https%3A%2F%2Fbenjeffrey%2Ecom%2Fposts%2Fpandoc%2Dsyntax%2Dhighlighting%2Dcss%0A%20%20%20Colours%20from%20https%3A%2F%2Fgist%2Egithub%2Ecom%2Frobsimmons%2F1172277%20%2A%2F%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Keyword%20%2A%2F%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%2F%2A%20DataType%20%2A%2F%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%2F%2A%20DecVal%20%28decimal%20values%29%20%2A%2F%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20BaseN%20%2A%2F%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Float%20%2A%2F%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Char%20%2A%2F%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20String%20%2A%2F%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%2F%2A%20Comment%20%2A%2F%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%2F%2A%20OtherToken%20%2A%2F%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20AlertToken%20%2A%2F%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Function%20calls%20%2A%2F%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%2F%2A%20ErrorTok%20%2A%2F%0A%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Meta-analysis</h1>
<h4 class="author"><em>Alencar Xavier and Shizhong Xu</em></h4>
<h4 class="date"><em>2016-09-14</em></h4>



<div id="meta-analysis-of-multiple-populations-along-with-gxe" class="section level1">
<h1>Meta-analysis of multiple populations along with GxE</h1>
<div id="genome-wide-association" class="section level2">
<h2>Genome-wide association</h2>
<p>Let the dataset have <span class="math">\(n\)</span> observations, <span class="math">\(f\)</span> subpopulations, <span class="math">\(m\)</span> markers and <span class="math">\(e\)</span> environments. The method starts from the genome-wide association analysis in the <span class="math">\(j^{th}\)</span> environment, following the alternative model <span class="math">\(y_j = \mu_j + Z\hat{\gamma}_i + \xi_j + \epsilon_j\)</span>. Where <span class="math">\(y\)</span> is the vector corresponding to the response variable, <span class="math">\(\mu\)</span> is the intercept, <span class="math">\(Z\)</span> is a <span class="math">\(n\times f\)</span> incidence matrix indicating the haplotype of the maker under evaluation, <span class="math">\(\hat{\gamma}_i\)</span> is a vector of the allele effects of the <span class="math">\(i^{th}\)</span> marker of <span class="math">\(f\)</span> subpopulations, <span class="math">\(\xi\)</span> is a vector of length <span class="math">\(n\)</span> corresponding to the polygenic term and <span class="math">\(\epsilon\)</span> is a vector of residuals with length <span class="math">\(n\)</span>.</p>
</div>
<div id="meta-analysis" class="section level2">
<h2>Meta-analysis</h2>
<p>For each <span class="math">\(i\)</span> marker, the meta-analysis is based upon the concept of sufficient statistics, assuming that environments are independent and all information in each environment can be expressed by the allele effects <span class="math">\(\hat{\gamma}_i\)</span> and the observed residual matrix <span class="math">\(R_i\)</span> obtained from the association analysis. The meta-analysis attempt to verify whether the genetic (<span class="math">\(G\)</span>) and environmental (<span class="math">\(E\)</span>) components of <span class="math">\(\gamma\)</span> differ from zero and, in addition, to verify the existance of <span class="math">\(G\times E\)</span> component. In this step, the set of <span class="math">\(\hat{\gamma}\)</span> from the association analyses becomes the response variable, a vector with length <span class="math">\(e\times f\)</span>. For the <span class="math">\(i^{th}\)</span> marker, variance components are obtained from the following random model: <span class="math">\(\gamma_i = \mu_i + Z\alpha_i + W\beta_i + H\delta_i + e_i\)</span>. Where <span class="math">\(\mu_i\)</span> is the intercept, <span class="math">\(Z\)</span> is a <span class="math">\(ef\times f\)</span> incidence matrix indicating the allele source, <span class="math">\(\alpha_i\)</span> is the genetic effect associated to the marker, <span class="math">\(W\)</span> is a <span class="math">\(ef\times e\)</span> incidence matrix indicating the environmental factor, <span class="math">\(\beta_i\)</span> is the coefficient associated to each environment, <span class="math">\(H\)</span> is the incidence matrix of genotype by environment interaction, <span class="math">\(\delta_i\)</span> is the coefficient associated to the <span class="math">\(G\times E\)</span> term, and <span class="math">\(e_i\)</span> is the vector of residuals with a known residual covariance matrix <span class="math">\(R\)</span>, a block diagonal matrix <span class="math">\(ef\times ef\)</span>.</p>
</div>
<div id="ammi-term" class="section level2">
<h2>AMMI term</h2>
<p>The <span class="math">\(G\times E\)</span> term might saturate the model once each regression coefficient <span class="math">\(\gamma\)</span> is observed as an unreplicated combination of genotype and environment. The saturation does not occur because the residuals are not independent and the structure is known. Yet, there exist an alternative reparameterization of this term: the additive main effect and multiplicative interaction (AMMI) term. The AMMI term works as follows: Suppose that the analysis are being performed in a dataset with <span class="math">\(f=5\)</span> subpopulations and <span class="math">\(e=4\)</span> environments. Once <span class="math">\(\gamma\)</span> has been estimated from the association analysis (step 1) and variance components of the genetic and environmental have been estimated with meta-analysis (step 2), then one can build the following <span class="math">\(E\)</span> matrix of residuals that also contains the higher-order interaction term:</p>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">E1</th>
<th align="left">E2</th>
<th align="left">E3</th>
<th align="left">E4</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">G1</td>
<td align="left"><span class="math">\(\varepsilon_{11}\)</span></td>
<td align="left"><span class="math">\(\varepsilon_{21}\)</span></td>
<td align="left"><span class="math">\(\varepsilon_{31}\)</span></td>
<td align="left"><span class="math">\(\varepsilon_{41}\)</span></td>
</tr>
<tr class="even">
<td align="left">G2</td>
<td align="left"><span class="math">\(\varepsilon_{12}\)</span></td>
<td align="left"><span class="math">\(\varepsilon_{22}\)</span></td>
<td align="left"><span class="math">\(\varepsilon_{32}\)</span></td>
<td align="left"><span class="math">\(\varepsilon_{42}\)</span></td>
</tr>
<tr class="odd">
<td align="left">G3</td>
<td align="left"><span class="math">\(\varepsilon_{13}\)</span></td>
<td align="left"><span class="math">\(\varepsilon_{23}\)</span></td>
<td align="left"><span class="math">\(\varepsilon_{33}\)</span></td>
<td align="left"><span class="math">\(\varepsilon_{43}\)</span></td>
</tr>
<tr class="even">
<td align="left">G4</td>
<td align="left"><span class="math">\(\varepsilon_{14}\)</span></td>
<td align="left"><span class="math">\(\varepsilon_{24}\)</span></td>
<td align="left"><span class="math">\(\varepsilon_{34}\)</span></td>
<td align="left"><span class="math">\(\varepsilon_{44}\)</span></td>
</tr>
<tr class="odd">
<td align="left">G5</td>
<td align="left"><span class="math">\(\varepsilon_{15}\)</span></td>
<td align="left"><span class="math">\(\varepsilon_{25}\)</span></td>
<td align="left"><span class="math">\(\varepsilon_{35}\)</span></td>
<td align="left"><span class="math">\(\varepsilon_{45}\)</span></td>
</tr>
</tbody>
</table>
<p>The AMMI term is extracted from the singlar-value decomposition (SVD). The SVD procedure is commonly used for the extraction of signals from non-square matrices. The decomposition is <span class="math">\(E = UDV\)</span>. Where, <span class="math">\(U\)</span> is a <span class="math">\(e\times e\)</span> matrix, <span class="math">\(D\)</span> is a <span class="math">\(e\times f\)</span> retangular diagonal matrix, and <span class="math">\(V\)</span> is a <span class="math">\(f\times f\)</span> matrix. In analogy to the Eigendecomposition, <span class="math">\(U\)</span> and <span class="math">\(V\)</span> represent Eigenvectors while <span class="math">\(D\)</span> are Eigenvalues. Likewise, a small fraction of principal components contain the most amount of information to recontruct the original matrix. Suppose one recontructs <span class="math">\(E\)</span> using the first <span class="math">\(p=2\)</span> principal components:</p>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">E1</th>
<th align="left">E2</th>
<th align="left">E3</th>
<th align="left">E4</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">G1</td>
<td align="left"><span class="math">\(q_{11}\)</span></td>
<td align="left"><span class="math">\(q_{21}\)</span></td>
<td align="left"><span class="math">\(q_{31}\)</span></td>
<td align="left"><span class="math">\(q_{41}\)</span></td>
</tr>
<tr class="even">
<td align="left">G2</td>
<td align="left"><span class="math">\(q_{12}\)</span></td>
<td align="left"><span class="math">\(q_{22}\)</span></td>
<td align="left"><span class="math">\(q_{32}\)</span></td>
<td align="left"><span class="math">\(q_{42}\)</span></td>
</tr>
<tr class="odd">
<td align="left">G3</td>
<td align="left"><span class="math">\(q_{13}\)</span></td>
<td align="left"><span class="math">\(q_{23}\)</span></td>
<td align="left"><span class="math">\(q_{33}\)</span></td>
<td align="left"><span class="math">\(q_{43}\)</span></td>
</tr>
<tr class="even">
<td align="left">G4</td>
<td align="left"><span class="math">\(q_{14}\)</span></td>
<td align="left"><span class="math">\(q_{24}\)</span></td>
<td align="left"><span class="math">\(q_{34}\)</span></td>
<td align="left"><span class="math">\(q_{44}\)</span></td>
</tr>
<tr class="odd">
<td align="left">G5</td>
<td align="left"><span class="math">\(q_{15}\)</span></td>
<td align="left"><span class="math">\(q_{25}\)</span></td>
<td align="left"><span class="math">\(q_{35}\)</span></td>
<td align="left"><span class="math">\(q_{45}\)</span></td>
</tr>
</tbody>
</table>
<p>The matrix above can be rearranged as a vector, and be included into the model of meta-analysis replacing the current <span class="math">\(G\times E\)</span> term. Thus, the model for the <span class="math">\(i^{th}\)</span> marker can be also expressed as <span class="math">\(\gamma_i = \mu_i + Z\alpha_i + W\beta_i + Q\tau_i + e_i\)</span>.</p>
</div>
<div id="hypothesis-testing" class="section level2">
<h2>Hypothesis testing</h2>
<p>The log-likelihood of the model is, therefore, <span class="math">\(L(\mu,\sigma^2_\alpha,\sigma^2_\beta,\sigma^2_\tau) = -0.5( log|V|+(y-\mu)^TV^{-1}(y-\mu) )\)</span>, where the variance is expressed as <span class="math">\(\sigma^2_\gamma = V = ZZ^T\sigma^2_\alpha + WW^T\sigma^2_\beta + QQ^T\sigma^2_\tau + R\)</span> and the log-likelihood of the model above is tested agaist <span class="math">\(L(\mu=\sigma^2_\alpha=\sigma^2_\beta=\sigma^2_\tau=0)\)</span>, providing the evidence that at least one of the coefficients (intercept and variance components) is not null. Thus, <span class="math">\(LRT = -2(L_{\mu,\sigma^2_\alpha,\sigma^2_\beta,\sigma^2_\tau}-L_{0,0,0,0})\)</span>.</p>
</div>
<div id="woodburys-matrix-identities" class="section level2">
<h2>Woodbury’s matrix identities</h2>
<p>The computational burden associated to the analysis above is originated from the determinant and inversion of the covariance matrix <span class="math">\(V\)</span>, a square matrix with <span class="math">\(ef\)</span> rows and columns. Let <span class="math">\(X=[Z\sigma_\alpha || W\sigma_\beta || Q\sigma_\tau]\)</span>, such that <span class="math">\(V = XX^T+R\)</span>. Using the Woodbury’s matrix identities, we have <span class="math">\(V^{-1} = R^{-1}-R^{-1}X(X^TR^{-1}X+I)^{-1}X^TR^{-1}\)</span> and <span class="math">\(|V| = |X^TR^{-1}X+I||R|\)</span>, where the square matrix to be inverted has dimension <span class="math">\(e+f+p\)</span>. For the example, in the analysis of a dataset with <span class="math">\(e=18\)</span> environments, <span class="math">\(f=41\)</span> subpopulations and using <span class="math">\(p=2\)</span> principal components for the <span class="math">\(G\times E\)</span> term, we invert a square matrix with dimension <span class="math">\(18+41+2=61\)</span> rows and columns intead of a matrix with <span class="math">\(18\times 41=738\)</span> rows and columns.</p>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
